src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
+
+	m_devices.erase(iter);
 }
 
 void GlobalShortcutWin::timeTicked() {
-	if (Global::get().mw->uiNewHardware != uiHardwareDevices) {
-		uiHardwareDevices = Global::get().mw->uiNewHardware;
-
-		XInputCheck_ClearDeviceCache();
-#ifdef USE_XBOXINPUT
-		nxboxinput = 0;
-#endif
-		pDI->EnumDevices(DI8DEVCLASS_ALL, EnumDevicesCB, static_cast< void * >(this), DIEDFL_ATTACHEDONLY);
-	}
-
-	if (bNeedRemap)
+	if (bNeedRemap) {
 		remap();
-
-	foreach (InputDevice *id, qhInputDevices) {
-		DIDEVICEOBJECTDATA rgdod[DX_SAMPLE_BUFFER_SIZE];
-		DWORD dwItems = DX_SAMPLE_BUFFER_SIZE;
-		HRESULT hr;
-
-		hr = id->pDID->Acquire();
-
-		switch (hr) {
-			case DI_OK:
-			case S_FALSE:
-				break;
-			case DIERR_UNPLUGGED:
-			case DIERR_GENERIC:
-				qWarning("Removing device %s", qPrintable(QUuid(id->guid).toString()));
-				id->pDID->Release();
-				qhInputDevices.remove(id->guid);
-				delete id;
-				return;
-			case DIERR_OTHERAPPHASPRIO:
-				continue;
-			default:
-				break;
-		}
-
-		{
-			QElapsedTimer timer;
-			timer.start();
-
-			id->pDID->Poll();
-
-			// If a call to Poll takes more than
-			// a second, warn the user that they
-			// might have a misbehaving device.
-			if (timer.elapsed() > 1000) {
-				qWarning("GlobalShortcut_win: Poll() for device %s took %li msec. This is abnormal, the device is "
-						 "possibly misbehavinGlobal::get()...",
-						 qPrintable(QUuid(id->guid).toString()), static_cast< long >(timer.elapsed()));
-			}
-		}
-
-		hr = id->pDID->GetDeviceData(sizeof(DIDEVICEOBJECTDATA), rgdod, &dwItems, 0);
-		if (HAS_FAILED(hr))
-			continue;
-
-		if (dwItems <= 0)
-			continue;
-
-		for (DWORD j = 0; j < dwItems; j++) {
-			QList< QVariant > ql;
-
-			quint32 uiType = id->qhOfsToType.value(rgdod[j].dwOfs);
-			ql << uiType;
-			ql << id->vguid;
-			handleButton(ql, rgdod[j].dwData & 0x80);
-		}
 	}
-
-#ifdef USE_GKEY
-	if (Global::get().s.bEnableGKey && gkey && gkey->isValid()) {
-		for (int button = GKEY_MIN_MOUSE_BUTTON; button <= GKEY_MAX_MOUSE_BUTTON; button++) {
-			QList< QVariant > ql;
-			ql << button;
-			ql << GKeyLibrary::quMouse;
-			handleButton(ql, gkey->isMouseButtonPressed(button));
-		}
-		for (int mode = GKEY_MIN_KEYBOARD_MODE; mode <= GKEY_MAX_KEYBOARD_MODE; mode++) {
-			for (int key = GKEY_MIN_KEYBOARD_BUTTON; key <= GKEY_MAX_KEYBOARD_BUTTON; key++) {
-				QList< QVariant > ql;
-				// Store the key and mode in one int
-				// bit 0..15: mode, bit 16..31: key
-				ql << (key | (mode << 16));
-				ql << GKeyLibrary::quKeyboard;
-				handleButton(ql, gkey->isKeyboardGkeyPressed(key, mode));
-			}
-		}
-	}
-#endif
-
 #ifdef USE_XBOXINPUT
-	if (Global::get().s.bEnableXboxInput && xboxinput && xboxinput->isValid() && nxboxinput > 0) {
-		XboxInputState state;
-		for (uint32_t i = 0; i < XBOXINPUT_MAX_DEVICES; i++) {
-			if (xboxinput->GetState(i, &state) == 0) {
+	if (m_xinput && m_xinputDevices > 0) {
+		for (uint8_t i = 0; i < XBOXINPUT_MAX_DEVICES; ++i) {
```

Comment:
```suggestion
		static_assert(std::numeric_limits< uint8_t >::max() >= XBOXINPUT_MAX_DEVICES, "Higher XBoxInput max devices number than anticipated.");
		for (uint8_t i = 0; i < XBOXINPUT_MAX_DEVICES; ++i) {
```
Just to be on the safe side at all times and to erase any doubts a programmer might have about the assumption about the size of `XBOXINPUT_MAX_DEVICES`

Note: This would require the include of `<limits>`

Reply:
Right, I was concerned about the same thing.